sdk/core/core-client-rest/src/apiVersionPolicy.ts

```diff
@@ -17,7 +17,10 @@ export function apiVersionPolicy(options: ClientOptions): PipelinePolicy {
     sendRequest: (req, next) => {
       if (options.apiVersion) {
         const url = new URL(req.url);
-        url.searchParams.append("api-version", options.apiVersion);
+        // Do not append customized api-version if url already has one e.g LRO/paging URL returned by the service
+        if (!url.searchParams.has("api-version")) {
```

Comment:
@lmazuel, @mikekistler  what's the consensus on appending apiVersion on LRO and Paginated calls? Does the api-version already in the location/continuation url wins? Or do we always make the client api-view override the set by the service?

I've always thought of continuation tokens and location URLs as Opaque values that happen to be URLs most of the time. Given this, my feeling is that we shouldn't try to override what's in there originally so having the service set api-version makes sense to me.

/cc: @bterlson, @jeremymeng, @mpodwysocki

Reply:
The [Azure SDK Guidelines](https://azure.github.io/azure-sdk/general_design.html#general-service-apiversion-7) say:
> ✅ DO add or replace the api-version query parameter on any URI returned by the service e.g., Operation-Location, next page links, etc., with the service version passed configured on the client.

Given that this is in the official guidelines I assume that there is agreement among the architects on this, or at least there was at the time the guidelines were written / adopted.

Reply:
so the correct fix here would be to remove the one in the url and overwrite it?

Reply:
Yes, the api-version sent in requests should always be the api-version of the client.

Reply:
@mikekistler  Could you suggest on how to handle or be backward compatible with below case we met in containerservice?

## Case: The server reports 404 after replacing the existing api-version with client one

1. The client sends the initial request to create `PUT Microsoft.ContainerService/managedClusters` with the client api version `2022-05-02-preview` to initiate the long-running operation.

```yaml
PUT https://management.azure.com/.../managedClusters/myreourcexyz?api-version=2022-05-02-preview
```
2. The resource sends a response to client with a 201-Created status code and its header `Azure-AsyncOperation` links a polling URL with old api-version with a "non-terminal" status

```yaml
https://management.azure.com/.../operations/{operationId}?api-version=2017-08-31
```
3. The client replaces the old version with the client one and then issues a GET request to operation status
```yaml
GET https://management.azure.com/.../operations/{operationId}?api-version=2022-05-02-preview
```
4. The resource responds with a 404 status code with below error message
```
"error": {
    "code": "InvalidResourceType",
    "message": "The resource type locations/operations could not be found in the namespace Microsoft.ContainerService for api version 2022-05-02-preview. The supported api-versions are 2016-03-30,2017-01-31,2017-08-31,2018-10-31,2019-11-27,2016-09-30,2017-07-01,2015-11-01-preview."
 }
```

Reply:
This looks like a bug -- in fact two bugs -- in the service implementation.

1. The service should not be returning an older api-version in Operation-Location than the version on the initial request, and
2. The service should not reject a request to the Operation-Location URL with the newer api-version.

Can you start a discussion with the Container Service team on these?

Reply:
Sure, I init this thread and included you.

Reply:
@mikekistler I'd like to confirm one thing with you that is `client api-version`. Currently we have two kinds of api-version in client.

- One is the default api-version derived from swagger definition
- The other is configured by the SDK users as client-level parameter

They are both `client api-version`, Right?

Reply:
I would say that only the value passed by the SDK user is the "client api-version", and only when the user did not supply a value -- taking the default -- that the default api-version becomes the client api-version.

Reply:
Just for your information.
/cc @qiaozha @joheredi @lirenhe @tadelesh

Reply:
The "default" is the default version for the client. Every request sent from that client should have the api-version of the client.

I don't know the details of the JS design but if it has no concept of an api-version for the client it will need to determine the correct API version to substitute into the URL in these cases.  A good choice would be the api-version for the operation that initiated the LRO or paging.

Reply:
From customers perspective, they are not aware of either those LRO polling apis or those nextLinks from pagination. And we don't even have swagger defined for those apis. if we are truly make this as a standard, how do we make sure service teams are really following our standard without calling each apis ? Otherwise, we may never found out the problem like ContainerService's case.

Reply:
Getting service teams to do the right thing has always been a struggle. We do our best to make clear REST API guidelines and enforce these as best we can in REST API reviews. We are also working on tooling to verify service implementations follow our guidelines -- this will also help some. I don't know that we will ever eliminate cases like this but we are trying to minimize them.